src/main.rs

```diff
@@ -0,0 +1,289 @@
+//! STAR Randomness web service
+
+use axum::extract::{Json, State};
+use axum ::http::StatusCode;
+use axum::{routing::get, routing::post, Router};
+use base64::prelude::{Engine as _, BASE64_STANDARD as BASE64};
+use serde::{Deserialize, Serialize};
+use tracing::{debug, info};
+
+use ppoprf::ppoprf;
+use std::sync::{Arc, RwLock};
+
+
+/// Internal state of the OPRF service
+struct OPRFServer {
+    /// oprf implementation
+    server: ppoprf::Server,
+    /// currently-valid randomness epoch
+    epoch: u8,
+}
+
+/// Shareable wrapper around the server state
+type OPRFState = Arc<RwLock<OPRFServer>>;
+/// Represents an contiguous range of configured randomness epochs
+type EpochRange = std::ops::Range<u8>;
+
+impl OPRFServer {
+    /// Initialize a new OPRFServer state supporting the given list of epochs
+    fn new(epochs: &EpochRange) -> Result<Self, ppoprf::PPRFError> {
+        // ppoprf wants a vector, so generate one from our range.
+        let epochs: Vec<u8> = epochs.to_owned().collect();
+        let epoch = epochs[0];
+        let server = ppoprf::Server::new(epochs)?;
+        Ok(OPRFServer{ server, epoch })
+    }
+}
+
+
+/// Request format for the randomness endpoint
+#[derive(Deserialize, Debug)]
+struct RandomnessRequest {
+    /// Array of points to evaluate
+    /// Should be base64-encoded compressed Ristretto curve points
+    points: Vec<String>,
+    /// Optional request for evaluation within a specific epoch
+    epoch: Option<u8>,
+}
+
+/// Response format for the randomness endpoint
+#[derive(Serialize, Debug)]
+struct RandomnessResponse {
+    /// Resulting points from the OPRF valuation
+    /// Should be base64-encode compressed points in one-to-one
+    /// correspondence with the request points.
+    points: Vec<String>,
+    /// Randomness epoch used in the evaluation.
+    epoch: u8,
+}
+
+/// Maximum number of points acceptable in a single request
+const MAX_POINTS: usize = 1024;
+
+/// Response format for the info endpoint
+/// Rename fields to match the earlier go implementation.
+#[derive(Serialize, Debug)]
+struct InfoResponse {
+    #[serde(rename = "publicKey")]
+    public_key: String,
+    #[serde(rename = "currentEpoch")]
+    current_epoch: u8,
+    #[serde(rename = "nextEpochTime")]
+    next_epoch_time: String,
+    #[serde(rename = "maxPoints")]
+    max_points: usize,
+}
+
+/// Response returned to report error conditions
+#[derive(Serialize, Debug)]
+struct ErrorResponse {
+    /// Human-readable description of the error
+    message: String,
+}
+
+/// Server error conditions
+#[derive(Debug)]
+enum Error {
```

Comment:
I suggest using the `thiserror` crate for this enum. It will implement the `Error` and `Display` trait for you, as well as the `From` trait for the `Base64` and `Oprf` errors, so you don't have to call `map_err` yourself. This will make the `IntoResponse` impl leaner, and allow you to use the Error enum for other things such as logging.

Reply:
Hmm. It's certainly more flexible and keeps the enum and messages together. But this code is supposed to be audited by users, so I was trying to keep things shallow. There are only a couple of handlers here, so I think explicit might be better than implicit `map_err` through the `From` trait.

How strongly do you feel about it, @DJAndries?

Using `thiserror` adds about 4k to the binary size, which isn't significant.

Reply:
I still think it makes sense to implement the error trait for the following reasons:

- it's the proper pattern for error types in Rust
- makes our lives easier by being able to use the `?` syntax for automatically converting any covered error to our error type
  - i understand there's only a couple places where we use `map_err`, but in the future the randomness server could become more advanced and it could be nice to not have to explicitly call `map_err`.
- the semantics for getting details from the error are better
  - the `source` method can be used to retrieve the lower-level error, which can be noted using the `#[source]` macro provided by the `thiserror` crate
  - the `Display` trait makes it easy to get a useful, human-readable error message from the error; if we wanted to log the error via `tracing`, or present the error in some other place that isn't the HTTP response, it will be very easy to do so; more versatile
- if we have multiple error types in the future (which i admit is probably unlikely, so this isn't my strongest point), we can return boxed errors easily via `Result<T, Box<dyn Error>>`

From an auditability perspective, I suspect that it won't be much of an issue. The `thiserror` crate is pretty well known, and the derive macros are easy to read. 

An alternative is to use `anyhow` which doesn't require you to create a custom Error enum, and converts any error into an `anyhow::Error` which implements `Error`-like methods and `Display`. It's a nice crate, but doesn't allow you to define error messages for each error in a declarative way like `thiserror`. A lazy person's crate, hence why I mention it :)

Reply:
yeah, i see how implementing `From` for a generic type could be tricky. I think specifying `PoisonError<OPRFState>` would be the only way to do this via thiserror.